
# Analyzing Hacker News Posts Data

Hacker News is a platform, extremely popular in technology and startup circles, where users share their posts and receives some votes and comments. We would be analyzing these posts along with some additional data like date of creation, upvotes, comments and so on, to derive some useful insights. The [data set]('https://www.kaggle.com/hacker-news/hacker-news-posts') used has been cut down from almost 300,000 rows to approximately 20,000 rows for our analysis purposes by removing those posts' data which have incomplete information such as zero comments or upvotes.

The columns of the data set are as follows:

| Column  |About |
|:--------|:------|
|id | the unique identifier from Hacker News for the post|
|title | the title of the post|
|url | the URL that the posts links to, if the post has a URL|
|num_points| the number of points the post acquired, calculated as the total number of upvotes minus the total number of downvotes|
|num_comments | the number of comments on the post|
|author | the username of the person who submitted the post|
|created_at | the date and time of the post's submission|

We are specifically interested in posts beggining with `Ask HN` or `Show HN`. `Ask HN` are the posts in which users asks a question to the other users on Hacker News and `Show HN` are the posts in which users showcase something of their own or just anything interesting. 

We would be comparing these two posts to determine the following:
* Do `Ask HN` or `Show HN` receive more comments on average?
* Do posts created at a certain time receive more comments on average?


__Importing the required Hacker News data set__

Let's start by importing the data set, `hacker_news.csv` as a list of lists and storing it in `hn`.

In [1]:
# Importing 'reader' from 'csv' module
from csv import reader

# Opening the file and creating a 'reader' object
opened_file = open('hacker_news.csv')
read_file = reader(opened_file)

hn = list(read_file)
hn[:5]

[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at'],
 ['12224879',
  'Interactive Dynamic Video',
  'http://www.interactivedynamicvideo.com/',
  '386',
  '52',
  'ne0phyte',
  '8/4/2016 11:52'],
 ['10975351',
  'How to Use Open Source and Shut the Fuck Up at the Same Time',
  'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/',
  '39',
  '10',
  'josep2',
  '1/26/2016 19:30'],
 ['11964716',
  "Florida DJs May Face Felony for April Fools' Water Joke",
  'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/',
  '2',
  '1',
  'vezycash',
  '6/23/2016 22:20'],
 ['11919867',
  'Technology ventures: From Idea to Enterprise',
  'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429',
  '3',
  '1',
  'hswarna',
  '6/17/2016 0:01']]

Since, the above data set contains the header columns list as the first row, we will remove it from the set and store it in the `headers`.

In [2]:
# Storing headers of data in a variable
headers = hn[0]

# Assigning the data set with minus header row
hn = hn[1:]

# Inspecting the data set
print(headers)
print('\n')
print(hn[:5])

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']


[['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01'], ['10301696', 'Note by Note: The Making of Steinway L1037 (2007)', 'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0', '8', '2', 'walterbell', '9/30/2015 4:12']]


__Data cleaning__

Now that we have the data set, let's start preparing it for analysis. Since, we require posts that either starts with `Ask HN` or `Show HN`, we will isolate this data in other lists. We will use a string method `string.startswith('word')` to check whether a string starts with provided phrase or word. It will return `True` if it does, otherwise `False`.

In [3]:
# Defining some empty lists
ask_posts = []
show_posts = []
other_posts = []

# Looping over hn 
for row in hn:
    
    # Extracting the title and converting it to lower case
    title = row[1].lower()
    
    # Checking whether the title starts with 'ask hn' or 'show hn' 
    # and appending in respective lists.
    if title.startswith('ask hn'):
        ask_posts.append(row)
    elif title.startswith('show hn'):
        show_posts.append(row)
    else:
        other_posts.append(row)

print('Ask HN posts:',len(ask_posts))
print(ask_posts[:2])
print('\n')
print('Show HN posts:',len(show_posts))
print(show_posts[:2])
print('\n')
print('Other posts:',len(other_posts))
print(other_posts[:2])

Ask HN posts: 1744
[['12296411', 'Ask HN: How to improve my personal website?', '', '2', '6', 'ahmedbaracat', '8/16/2016 9:55'], ['10610020', 'Ask HN: Am I the only one outraged by Twitter shutting down share counts?', '', '28', '29', 'tkfx', '11/22/2015 13:43']]


Show HN posts: 1162
[['10627194', 'Show HN: Wio Link  ESP8266 Based Web of Things Hardware Development Platform', 'https://iot.seeed.cc', '26', '22', 'kfihihc', '11/25/2015 14:03'], ['10646440', 'Show HN: Something pointless I made', 'http://dn.ht/picklecat/', '747', '102', 'dhotson', '11/29/2015 22:46']]


Other posts: 17194
[['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30']]


__Finding the average number of comments on `Ask HN` and `Show HN` posts__

Now that we have different data for the `Ask HN` and `Show HN`, we can now start analysing data. We will determine whether `Ask HN` or `Show HN` posts receive more comments.

In [4]:
# Investigating the headers' data types
for column in headers:
    print(column,':',type(column))

# Finding the average number of comments for:
# 1) 'Ask HN' posts
total_ask_comments = 0
for row in ask_posts:
    num_comment = int(row[4])
    total_ask_comments += num_comment

print('\n')    
avg_ask_comments = total_ask_comments / len(ask_posts)
print('Average Ask HN comments : {:.2f}'.format(avg_ask_comments))

# 2) 'Show HN' posts
total_show_comments = 0
for row in show_posts:
    num_comment = int(row[4])
    total_show_comments += num_comment
    
avg_show_comments = total_show_comments / len(show_posts)
print('Average Show HN comments : {:.2f}'.format(avg_show_comments))

id : <class 'str'>
title : <class 'str'>
url : <class 'str'>
num_points : <class 'str'>
num_comments : <class 'str'>
author : <class 'str'>
created_at : <class 'str'>


Average Ask HN comments : 14.04
Average Show HN comments : 10.32


As it is evident from the above output, the average number of comments on `Ask HN` are greater than the `Show HN` posts. The reason could be because of the nature of `Ask HN` post which is basically a question being asked and thus engross much more audience as compared to the the `Show HN` post which is more like a general discussion, statement or a fact.

__Determining the most popular post-creation time__

Since the `Ask HN` posts seems more popular, we would continue our analysis on those only. We will determine if `Ask HN` posts created at certain time are more likely to attract comments. We'll use the following steps to perform this analysis:
* Calculate the number of `Ask HN` posts created in each hour of the day.
* Calculate the average number of comments `Ask HN` posts receive by hour.

In [5]:
# Importing required 'datetime' module
import datetime as dt

# Initializing an empty list for date of post creation and number of comments
# on the post
result_list = []

# Looping over 'Ask HN' posts
for row in ask_posts:
    
    # Storing the date and comments in variables
    created_at = row[-1]
    num_comments = int(row[-3])
    
    # Appending the above data to the list created before
    result_list.append([created_at, num_comments])
print('Sample from result_list:',result_list[:2])    
print('\n')

# Initializing two empty dictionaries for creating a frequency table for number
# of posts created per given hour and number of comments on the posts at each
# hour
counts_by_hour = {}
comments_by_hour = {}
for row in result_list:
    
    # Creating a datetime object from the string datetime column
    date_parsed = dt.datetime.strptime(row[0],'%m/%d/%Y %H:%M') 
    
    # Formatting the datetime object to obtain hour as string
    hr=date_parsed.strftime('%H')
    
    # Checking whether the above 'hour' exists as a key in the dictionaries or not
    # and updating the both the dictionaries as follows:
    # if hr exists in counts_by_hr then:
    #    update the key-value pair of hr in both dictionaries by 1
    # else:
    #    create the key-value pair of hr in both dictionaries with initialization at 1
    if hr in counts_by_hour:
        counts_by_hour[hr] += 1
        comments_by_hour[hr] += row[1]
    else:
        counts_by_hour[hr] = 1
        comments_by_hour[hr] = row[1]
        
print('Posts per hour:',counts_by_hour)
print('\n')
print('Comments per hour:',comments_by_hour)

Sample from result_list: [['8/16/2016 9:55', 6], ['11/22/2015 13:43', 29]]


Posts per hour: {'09': 45, '13': 85, '10': 59, '14': 107, '16': 108, '23': 68, '12': 73, '17': 100, '15': 116, '21': 109, '20': 80, '02': 58, '18': 109, '03': 54, '05': 46, '19': 110, '01': 60, '22': 71, '08': 48, '04': 47, '00': 55, '06': 44, '07': 34, '11': 58}


Comments per hour: {'09': 251, '13': 1253, '10': 793, '14': 1416, '16': 1814, '23': 543, '12': 687, '17': 1146, '15': 4477, '21': 1745, '20': 1722, '02': 1381, '18': 1439, '03': 421, '05': 464, '19': 1188, '01': 683, '22': 479, '08': 492, '04': 337, '00': 447, '06': 397, '07': 267, '11': 641}


__Calculating the average number of comments per `Ask HN` post for each hour__

Now that we have the data of number of posts and comments per hour, we can determine the average number of comments on a post for a given hour.

In [6]:
# Initializing an empty list for storing average number of comments per post
# for each hour i.e., number of comments per post per hour / num of posts per 
# hour
avg_by_hour = []
for post in counts_by_hour:
    avg_comments_per_post = comments_by_hour[post] / counts_by_hour[post]
    
    # Appending the data, time and averaged comments to the above list
    avg_by_hour.append([post,avg_comments_per_post])
print(avg_by_hour)

[['09', 5.5777777777777775], ['13', 14.741176470588234], ['10', 13.440677966101696], ['14', 13.233644859813085], ['16', 16.796296296296298], ['23', 7.985294117647059], ['12', 9.41095890410959], ['17', 11.46], ['15', 38.5948275862069], ['21', 16.009174311926607], ['20', 21.525], ['02', 23.810344827586206], ['18', 13.20183486238532], ['03', 7.796296296296297], ['05', 10.08695652173913], ['19', 10.8], ['01', 11.383333333333333], ['22', 6.746478873239437], ['08', 10.25], ['04', 7.170212765957447], ['00', 8.127272727272727], ['06', 9.022727272727273], ['07', 7.852941176470588], ['11', 11.051724137931034]]


__Sorting the data for better accessibility and readability__

Now that we have the data, we can finally make a decision on the hour of the day on which the most comments were posted on a post. But the values are not easy to read since data is not sorted. Let's do so by swapping the values in each row of the above list and storing them in another list.

In [7]:
# Initializing an empty list for creating a swapped list
swap_avg_by_hour = []
for row in avg_by_hour:
    
    # Swapping and appending the values
    swap_avg_by_hour.append([row[1],row[0]])

print('Swapped list of data')    
for row in swap_avg_by_hour:
    print(row)
        
print('\n')

# Printing a sorted list of above data in descending order of average comments
sorted_swap = sorted(swap_avg_by_hour,reverse=True)
print('Sorting the above data')
for row in sorted_swap:
    print(row)

# Printing the top 5 values out of the above list
print('\n')
print('Top 5 hours for Ask HN posts')
for row in sorted_swap[:5]:
    hour = row[1]
    avg_comments = row[0]
    parsed_date = dt.datetime.strptime(hour,'%H')
    converted_date = parsed_date.strftime('%H:%M')
    
    print('{} EST: {:.2f} average comments per post'.format(converted_date,avg_comments))

Swapped list of data
[5.5777777777777775, '09']
[14.741176470588234, '13']
[13.440677966101696, '10']
[13.233644859813085, '14']
[16.796296296296298, '16']
[7.985294117647059, '23']
[9.41095890410959, '12']
[11.46, '17']
[38.5948275862069, '15']
[16.009174311926607, '21']
[21.525, '20']
[23.810344827586206, '02']
[13.20183486238532, '18']
[7.796296296296297, '03']
[10.08695652173913, '05']
[10.8, '19']
[11.383333333333333, '01']
[6.746478873239437, '22']
[10.25, '08']
[7.170212765957447, '04']
[8.127272727272727, '00']
[9.022727272727273, '06']
[7.852941176470588, '07']
[11.051724137931034, '11']


Sorting the above data
[38.5948275862069, '15']
[23.810344827586206, '02']
[21.525, '20']
[16.796296296296298, '16']
[16.009174311926607, '21']
[14.741176470588234, '13']
[13.440677966101696, '10']
[13.233644859813085, '14']
[13.20183486238532, '18']
[11.46, '17']
[11.383333333333333, '01']
[11.051724137931034, '11']
[10.8, '19']
[10.25, '08']
[10.08695652173913, '05']
[9.41095890410959, '12

The above results suggests that the time period from `15:00 to 16:00 EST` that is Eastern Standard Time, receives the highest number of comments on the `Ask HN` posts. To understand it in the terms of `IST` that is, `Indian Standard Time` let's do a conversion.

In [8]:
# Initializing an empty list for storing IST values
ist_data = []

# Importing 'pytz' module for defining the timezone of original values as 'EST'
import pytz
for row in sorted_swap:
    avg_comment = row[0]
    time = row[1]
    
    # Parsing the date from string as hour only 
    parse_time = dt.datetime.strptime(time,'%H')
    
    # Replacing the naive 'datetime' object with 'active' datetime object with
    # 'UST' timezone
    est_tz = parse_time.replace(tzinfo=pytz.timezone('US/Eastern'))
    
    # Converting the timezone: EST to IST
    changed_tz = est_tz.astimezone(pytz.timezone('Asia/Kolkata'))
    
    # Finally appending the data to the list as formatted string
    final_ist = changed_tz.strftime('%H:00')
    ist_data.append([avg_comment,final_ist])

# printing the top 5 and the least average number of comments posts
print('Top 5 hours for Ask HN post in Indian Standard Time')    
for row in ist_data[:5]:
        time_data = row[1]
        avg_comment = row[0]
        print('{} IST: {:.2f} average comments per post'.format(time_data,avg_comment))
print('Hour with least comments: {} IST with {:.2f} comments'.format(ist_data[-1][-1],ist_data[-1][0]))        

Top 5 hours for Ask HN post in Indian Standard Time
01:00 IST: 38.59 average comments per post
12:00 IST: 23.81 average comments per post
06:00 IST: 21.52 average comments per post
02:00 IST: 16.80 average comments per post
07:00 IST: 16.01 average comments per post
Hour with least comments: 19:00 IST with 5.58 comments


Now the data seems to be much better. It can be easily deduced that difference between the hour with maximum average number of comments and the minimum average number of comments comes at a percentage difference of approximately `30%`, which is quite significant. Hence, we can say that the timings of post creation does infact impacts the number of comments and thus audience engagement.


## Conclusion
In this project, we analyzed `Ask HN` posts and `Show HN` posts to determine which type of post and time receive the most comments on average. Based on our analysis, to maximize the amount of comments a post receives, we'd recommend the post be categorized as `Ask HN` post and created between `15:00 and 16:00 EST` `(1:00 and 2:00 IST)`

However, it should be noted that the data set we analyzed excluded posts without any comments. Given that, it's more accurate to say that of the posts that received comments, `Ask HN` posts received more comments on average and `Ask HN` posts created between `15:00 and 16:00 EST` `(1:00 and 2:00 IST)` received the most comments on average.

## Some more useful insights from the data set
1. __Determining if `Show HN` or `Ask HN` receives more points on average__

In [9]:
# Initializing two lists to store posts' titles and points
ask_points = []
show_points = []

# Appending required data
for row in ask_posts:
    ask_points.append([row[1],int(row[3])])
for row in show_posts:
    show_points.append([row[1],int(row[3])])

# Calculating average points for Ask HN posts    
summed = 0    
for row in ask_points:
    point = row[1]
    summed += point
avg_ask_points = summed / len(ask_points)

# Calculating average points for Show HN posts
summed = 0    
for row in show_points:
    point = row[1]
    summed += point
avg_show_points = summed / len(show_points)

print('Average \'Ask HN\' points : {:.2f}'.format(avg_ask_points))
print('Average \'Show HN\' points : {:.2f}'.format(avg_show_points))

Average 'Ask HN' points : 15.06
Average 'Show HN' points : 27.56


It's clear from the above output that the `Show HN` posts have greater points as compared to `Ask HN` posts.

2. __Determining if posts created at a certain time are more likely to receive more points__

Since, `Show HN` posts have more number of average points, we will restrict our analysis to these only. In order to find out whether the posts created at certain time receive more points, we will do the following:
* Determining the number of `Show HN` posts created at different hours.
* Determining the average number of points received on the `Show HN` posts at particular hour.

In [10]:
# Initializing a list for storing Show HN posts' creation time and points
mod_show_posts = []

# Looping over Show HN posts data set and collecting required columns
for row in show_posts:
    date_post = row[-1]
    points_post = int(row[3])
    
    # Parsing the date using datetime.strptime() method 
    date_post = dt.datetime.strptime(date_post,'%m/%d/%Y %H:%M')
    
    # Appending the data to the above list
    mod_show_posts.append([date_post,points_post])

# Checking the data set
print('Sample data:',mod_show_posts[:5])

print('\n')

# Initializing two empty dictionaries for creating frequency tables for number
# of posts created per hour and number of points received per hour per
show_by_hour = {}
points_by_hour = {}


for row in mod_show_posts:
    points = row[1]
    date = row[0]
    
    # Extracting the hour out of the datetime object
    hour = date.strftime('%H')
    
    # Checking if hour exists as a key in both dictionaries
    if hour in show_by_hour:
        show_by_hour[hour] += 1
        points_by_hour[hour] += points
    else:
        show_by_hour[hour] = 1
        points_by_hour[hour] = points
        
print('Posts per hour:',show_by_hour)
print('\n')
print('Points per hour:',points_by_hour)

    

Sample data: [[datetime.datetime(2015, 11, 25, 14, 3), 26], [datetime.datetime(2015, 11, 29, 22, 46), 747], [datetime.datetime(2016, 4, 28, 18, 5), 1], [datetime.datetime(2016, 7, 28, 7, 11), 3], [datetime.datetime(2016, 1, 9, 20, 45), 1]]


Posts per hour: {'14': 86, '22': 46, '18': 61, '07': 26, '20': 60, '05': 19, '16': 93, '19': 55, '15': 78, '03': 27, '17': 93, '06': 16, '02': 30, '13': 99, '08': 34, '21': 47, '04': 26, '11': 44, '12': 61, '23': 36, '09': 30, '01': 28, '10': 36, '00': 31}


Points per hour: {'14': 2187, '22': 1856, '18': 2215, '07': 494, '20': 1819, '05': 104, '16': 2634, '19': 1702, '15': 2228, '03': 679, '17': 2521, '06': 375, '02': 340, '13': 2438, '08': 519, '21': 866, '04': 386, '11': 1480, '12': 2543, '23': 1526, '09': 553, '01': 700, '10': 681, '00': 1173}


In [11]:
# Initializing a list for storing averaged points per post per hour
avg_points_post = []
for hour in show_by_hour:
    post_count = show_by_hour[hour]
    post_points = points_by_hour[hour]
    
    # Average points per post for every hour = no. of points per hour / no. of posts per hour
    avg_data =  post_points / post_count
    avg_points_post.append([hour,avg_data])

avg_points_post
    

[['14', 25.430232558139537],
 ['22', 40.34782608695652],
 ['18', 36.31147540983606],
 ['07', 19.0],
 ['20', 30.316666666666666],
 ['05', 5.473684210526316],
 ['16', 28.322580645161292],
 ['19', 30.945454545454545],
 ['15', 28.564102564102566],
 ['03', 25.14814814814815],
 ['17', 27.107526881720432],
 ['06', 23.4375],
 ['02', 11.333333333333334],
 ['13', 24.626262626262626],
 ['08', 15.264705882352942],
 ['21', 18.425531914893618],
 ['04', 14.846153846153847],
 ['11', 33.63636363636363],
 ['12', 41.68852459016394],
 ['23', 42.388888888888886],
 ['09', 18.433333333333334],
 ['01', 25.0],
 ['10', 18.916666666666668],
 ['00', 37.83870967741935]]

In [12]:
# Creating a list with swapped values for better readability    
swapped = []    
for row in avg_points_post:
    swapped.append([row[1],row[0]])

print('\n')

# Sorting values in descending order for better accessibility
sorted_swap = sorted(swapped,reverse=True)
sorted_swap

[[42.388888888888886, '23'],
 [41.68852459016394, '12'],
 [40.34782608695652, '22'],
 [37.83870967741935, '00'],
 [36.31147540983606, '18'],
 [33.63636363636363, '11'],
 [30.945454545454545, '19'],
 [30.316666666666666, '20'],
 [28.564102564102566, '15'],
 [28.322580645161292, '16'],
 [27.107526881720432, '17'],
 [25.430232558139537, '14'],
 [25.14814814814815, '03'],
 [25.0, '01'],
 [24.626262626262626, '13'],
 [23.4375, '06'],
 [19.0, '07'],
 [18.916666666666668, '10'],
 [18.433333333333334, '09'],
 [18.425531914893618, '21'],
 [15.264705882352942, '08'],
 [14.846153846153847, '04'],
 [11.333333333333334, '02'],
 [5.473684210526316, '05']]

In [13]:
# Printing the top 5 hours with most points per page
print('Top 5 hours for Show HN posts in Eastern Standard Time')
for row in sorted_swap[:5]:
    print('{}:00 EST: {:.2f} average points per post'.format(row[1],row[0]))

# Printing the hour with least points per page
print('Hour with least average points: {}:00 EST with {:.2f} points'.format(sorted_swap[-1][1],sorted_swap[-1][0]))    

Top 5 hours for Show HN posts in Eastern Standard Time
23:00 EST: 42.39 average points per post
12:00 EST: 41.69 average points per post
22:00 EST: 40.35 average points per post
00:00 EST: 37.84 average points per post
18:00 EST: 36.31 average points per post
Hour with least average points: 05:00 EST with 5.47 points


In [14]:
# Initializing a list to store IST converted values
ist_data_2 = []

# Importing 'pytz' module for defining the timezone of original values as 'EST'
import pytz
for row in sorted_swap:
    avg_points = row[0]
    time = row[1]
    
    # Parsing the date from string as hour only 
    parse_time = dt.datetime.strptime(time,'%H')
    
    # Replacing the naive 'datetime' object with 'active' datetime object with
    # 'UST' timezone
    est_tz = parse_time.replace(tzinfo=pytz.timezone('US/Eastern'))
    
    # Converting the timezone: EST to IST
    changed_tz = est_tz.astimezone(pytz.timezone('Asia/Kolkata'))
    
    # Finally appending the data to the list as formatted string
    final_ist = changed_tz.strftime('%H:00')
    ist_data_2.append([avg_points,final_ist])

# printing the top 5 and the least average number of comments posts
print('Top 5 hours for Show HN post in Indian Standard Time')    
for row in ist_data_2[:5]:
        time_data = row[1]
        avg_points = row[0]
        print('{} IST: {:.2f} average points per post'.format(time_data,avg_points))
print('Hour with least comments: {} IST with {:.2f} comments'.format(ist_data_2[-1][-1],ist_data_2[-1][0]))

Top 5 hours for Show HN post in Indian Standard Time
09:00 IST: 42.39 average points per post
22:00 IST: 41.69 average points per post
08:00 IST: 40.35 average points per post
10:00 IST: 37.84 average points per post
04:00 IST: 36.31 average points per post
Hour with least comments: 15:00 IST with 5.47 comments


It is clear from the above output that time between `23:00 and 24:00 EST` `(09:00 and 10:00 IST)` is the best to create a `Show HN` post and receive maximum points. Also, the percentage difference between the hour receiving top and least number of points is around `37%`, which is quite significant. Hence, the timing plays an important role in getting greater points on `Show HN` posts.

3. __Comparing the average number of comments and points for `Ask HN` and `Show HN` posts to all the other posts.__

In [15]:
# Initializing lists for storing dates with comments and points seperately
other_comments = []
other_points = []

# Looping over data set of 'other' posts
for row in other_posts:
    date = dt.datetime.strptime(row[-1],'%m/%d/%Y %H:%M')
    points = row[3]
    comments = row[4]
    
    # Appending data to the lists
    other_points.append([date,points])
    other_comments.append([date,comments])

print('Other posts points:',other_points[:5])
print('\n')
print('Other posts comments:',other_comments[:5])

print('\n')

# Calculating average points and comments for all the other posts
sum_points = 0
for row in other_points:
    points = int(row[1])
    sum_points += points
avg_points = sum_points/len(other_points)
print('Average points for other posts : {:.2f}'.format(avg_points))

sum_comments = 0
for row in other_comments:
    comments = int(row[1])
    sum_comments += comments
avg_comments = sum_comments / len(other_comments)
print('Average comments for other posts : {:.2f}'.format(avg_comments))

Other posts points: [[datetime.datetime(2016, 8, 4, 11, 52), '386'], [datetime.datetime(2016, 1, 26, 19, 30), '39'], [datetime.datetime(2016, 6, 23, 22, 20), '2'], [datetime.datetime(2016, 6, 17, 0, 1), '3'], [datetime.datetime(2015, 9, 30, 4, 12), '8']]


Other posts comments: [[datetime.datetime(2016, 8, 4, 11, 52), '52'], [datetime.datetime(2016, 1, 26, 19, 30), '10'], [datetime.datetime(2016, 6, 23, 22, 20), '1'], [datetime.datetime(2016, 6, 17, 0, 1), '1'], [datetime.datetime(2015, 9, 30, 4, 12), '2']]


Average points for other posts : 55.41
Average comments for other posts : 26.87


As it clear from above, the average points and comments for all other posts which excludes `Ask HN` and `Show HN` posts, seems to greater than their points and comments. 
* The average points for other posts exceeds that in case the of `Show HN` posts by almost `28%`, and
* The average comments for other posts exceeds that in case the of `Ask HN` posts by almost `13%`.

In [16]:
# Initializing dictionaries to make frequency tables for other posts, points
# and comments generated per hour
points_per_hour = {}
comments_per_hour = {}
posts_per_hour = {}

# Checking whether hour exists in points dictionary as a key-value pair
for row in other_points:
    hour = row[0].strftime('%H')
    points = int(row[1])
    if hour in posts_per_hour:
        posts_per_hour[hour] += 1
        points_per_hour[hour] += points
    else:
        posts_per_hour[hour] = 1
        points_per_hour[hour] = points

# Checking whether hour exists in comments dictionary as a key-value pair
for row in other_comments:
    hour = row[0].strftime('%H')
    comments = int(row[1])
    if hour in comments_per_hour:
        comments_per_hour[hour] += comments
    else:
        comments_per_hour[hour] = comments

print('Other posts per hour:',posts_per_hour)
print('\n')
print('Other points per hour:',points_per_hour)
print('\n')
print('Other comments per hour:',comments_per_hour)

Other posts per hour: {'11': 660, '19': 980, '22': 758, '00': 611, '04': 454, '09': 534, '16': 1101, '18': 1084, '10': 591, '12': 789, '20': 911, '03': 407, '17': 1169, '14': 958, '13': 918, '01': 500, '23': 674, '08': 496, '02': 441, '21': 874, '15': 1040, '06': 408, '05': 388, '07': 448}


Other points per hour: {'11': 37995, '19': 58811, '22': 38079, '00': 35718, '04': 22549, '09': 28802, '16': 59655, '18': 58459, '10': 35746, '12': 45287, '20': 41218, '03': 23167, '17': 67777, '14': 59191, '13': 57398, '01': 25303, '23': 35068, '08': 26830, '02': 25786, '21': 43149, '15': 62964, '06': 18864, '05': 19387, '07': 25461}


Other comments per hour: {'11': 19532, '19': 26167, '22': 17635, '00': 16544, '04': 10953, '09': 14732, '16': 27959, '18': 29186, '10': 15728, '12': 23944, '20': 21080, '03': 10918, '17': 32727, '14': 30973, '13': 28363, '01': 11536, '23': 16592, '08': 13405, '02': 12254, '21': 20635, '15': 30700, '06': 8714, '05': 9768, '07': 12010}


In [17]:
# Initializing lists to store averaged points and comments per hour per post
average_points = []
average_comments = []

for hour in posts_per_hour:
    
    # Average points per post per hour = points per hour / posts per hour
    # Average comments per post per hour = comments per hour / posts per hour
    avg_points_per_post = points_per_hour[hour] / posts_per_hour[hour]
    avg_comments_per_post = comments_per_hour[hour] / posts_per_hour[hour]
    
    average_points.append([hour,avg_points_per_post])
    average_comments.append([hour,avg_comments_per_post])
    
print('Average points per post per hour')
for row in average_points:
    print(row)
print('\n')
print('Average comments per post per hour')
for row in average_comments:
    print(row)

Average points per post per hour
['11', 57.56818181818182]
['19', 60.01122448979592]
['22', 50.236147757255935]
['00', 58.4582651391162]
['04', 49.66740088105727]
['09', 53.93632958801498]
['16', 54.182561307901906]
['18', 53.928966789667896]
['10', 60.4839255499154]
['12', 57.3979721166033]
['20', 45.24478594950604]
['03', 56.92137592137592]
['17', 57.97861420017109]
['14', 61.78601252609603]
['13', 62.525054466230934]
['01', 50.606]
['23', 52.02967359050445]
['08', 54.09274193548387]
['02', 58.471655328798185]
['21', 49.369565217391305]
['15', 60.542307692307695]
['06', 46.23529411764706]
['05', 49.96649484536083]
['07', 56.832589285714285]


Average comments per post per hour
['11', 29.593939393939394]
['19', 26.701020408163266]
['22', 23.265171503957784]
['00', 27.076923076923077]
['04', 24.125550660792953]
['09', 27.588014981273407]
['16', 25.394187102633968]
['18', 26.924354243542435]
['10', 26.612521150592215]
['12', 30.34727503168568]
['20', 23.13940724478595]
['03', 26.8255528

In [18]:
# Creating swapped lists for above data for better readability
swapped_points = []
swapped_comments = []

for row in average_points:
    swapped_points.append([row[1],row[0]])

print('\n')    
for row in average_comments:
    swapped_comments.append([row[1],row[0]])

print('Swapped points:')
for row in swapped_points:
    print(row)
print('\n')
print('Swapped comments:')    
for row in swapped_comments:
    print(row)    



Swapped points:
[57.56818181818182, '11']
[60.01122448979592, '19']
[50.236147757255935, '22']
[58.4582651391162, '00']
[49.66740088105727, '04']
[53.93632958801498, '09']
[54.182561307901906, '16']
[53.928966789667896, '18']
[60.4839255499154, '10']
[57.3979721166033, '12']
[45.24478594950604, '20']
[56.92137592137592, '03']
[57.97861420017109, '17']
[61.78601252609603, '14']
[62.525054466230934, '13']
[50.606, '01']
[52.02967359050445, '23']
[54.09274193548387, '08']
[58.471655328798185, '02']
[49.369565217391305, '21']
[60.542307692307695, '15']
[46.23529411764706, '06']
[49.96649484536083, '05']
[56.832589285714285, '07']


Swapped comments:
[29.593939393939394, '11']
[26.701020408163266, '19']
[23.265171503957784, '22']
[27.076923076923077, '00']
[24.125550660792953, '04']
[27.588014981273407, '09']
[25.394187102633968, '16']
[26.924354243542435, '18']
[26.612521150592215, '10']
[30.34727503168568, '12']
[23.13940724478595, '20']
[26.825552825552826, '03']
[27.99572284003422, '1

In [19]:
# Initializing lists to store the sorted data for better accessibility
sorted_points = sorted(swapped_points,reverse=True)
sorted_comments = sorted(swapped_comments,reverse=True)

print('Sorted average points per hour per post:')
for row in sorted_points:
    print(row)
print('\n')    
print('Sorted average comments per hour per post:')
for row in sorted_comments:
    print(row)    

Sorted average points per hour per post:
[62.525054466230934, '13']
[61.78601252609603, '14']
[60.542307692307695, '15']
[60.4839255499154, '10']
[60.01122448979592, '19']
[58.471655328798185, '02']
[58.4582651391162, '00']
[57.97861420017109, '17']
[57.56818181818182, '11']
[57.3979721166033, '12']
[56.92137592137592, '03']
[56.832589285714285, '07']
[54.182561307901906, '16']
[54.09274193548387, '08']
[53.93632958801498, '09']
[53.928966789667896, '18']
[52.02967359050445, '23']
[50.606, '01']
[50.236147757255935, '22']
[49.96649484536083, '05']
[49.66740088105727, '04']
[49.369565217391305, '21']
[46.23529411764706, '06']
[45.24478594950604, '20']


Sorted average comments per hour per post:
[32.33089770354906, '14']
[30.896514161220043, '13']
[30.34727503168568, '12']
[29.593939393939394, '11']
[29.51923076923077, '15']
[27.99572284003422, '17']
[27.786848072562357, '02']
[27.588014981273407, '09']
[27.076923076923077, '00']
[27.026209677419356, '08']
[26.924354243542435, '18']
[26

In [20]:
# Printing the top 5 hours with most points per page for points data above
print('Top 5 hours for maximum points on all other posts in Eastern Standard Time')
for row in sorted_points[:5]:
    print('{}:00 EST: {:.2f} average points per post'.format(row[1],row[0]))

# Printing the hour with least points per page
print('Hour with least average points: {}:00 EST with {:.2f} points'.format(sorted_points[-1][1],sorted_points[-1][0]))

print('\n')

# Printing the top 5 hours with most comments per page for comments data above
print('Top 5 hours for maximum comments on other posts in Eastern Standard Time')
for row in sorted_comments[:5]:
    print('{}:00 EST: {:.2f} average comments per post'.format(row[1],row[0]))

# Printing the hour with least points per page
print('Hour with least average comments: {}:00 EST with {:.2f} points'.format(sorted_comments[-1][1],sorted_comments[-1][0]))    

Top 5 hours for maximum points on all other posts in Eastern Standard Time
13:00 EST: 62.53 average points per post
14:00 EST: 61.79 average points per post
15:00 EST: 60.54 average points per post
10:00 EST: 60.48 average points per post
19:00 EST: 60.01 average points per post
Hour with least average points: 20:00 EST with 45.24 points


Top 5 hours for maximum comments on other posts in Eastern Standard Time
14:00 EST: 32.33 average comments per post
13:00 EST: 30.90 average comments per post
12:00 EST: 30.35 average comments per post
11:00 EST: 29.59 average comments per post
15:00 EST: 29.52 average comments per post
Hour with least average comments: 06:00 EST with 21.36 points


In [21]:
# Initializing a list to store IST converted values for points data above
ist_data_2 = []

# Importing 'pytz' module for defining the timezone of original values as 'EST'
import pytz
for row in sorted_points:
    avg_points = row[0]
    time = row[1]
    
    # Parsing the date from string as hour only 
    parse_time = dt.datetime.strptime(time,'%H')
    
    # Replacing the naive 'datetime' object with 'active' datetime object with
    # 'UST' timezone
    est_tz = parse_time.replace(tzinfo=pytz.timezone('US/Eastern'))
    
    # Converting the timezone: EST to IST
    changed_tz = est_tz.astimezone(pytz.timezone('Asia/Kolkata'))
    
    # Finally appending the data to the list as formatted string
    final_ist = changed_tz.strftime('%H:00')
    ist_data_2.append([avg_points,final_ist])

# printing the top 5 and the least average number of comments posts
print('Top 5 hours for maximum points on other posts in Indian Standard Time')
for row in ist_data_2[:5]:
        time_data = row[1]
        avg_points = row[0]
        print('{} IST: {:.2f} average points per post'.format(time_data,avg_points))
print('Hour with least points: {} IST with {:.2f} comments'.format(ist_data_2[-1][-1],ist_data_2[-1][0]))

print('\n')

# Initializing a list to store IST converted values for comments data above
ist_data_3 = []
for row in sorted_comments:
    avg_points = row[0]
    time = row[1]
    
    # Parsing the date from string as hour only 
    parse_time = dt.datetime.strptime(time,'%H')
    
    # Replacing the naive 'datetime' object with 'active' datetime object with
    # 'UST' timezone
    est_tz = parse_time.replace(tzinfo=pytz.timezone('US/Eastern'))
    
    # Converting the timezone: EST to IST
    changed_tz = est_tz.astimezone(pytz.timezone('Asia/Kolkata'))
    
    # Finally appending the data to the list as formatted string
    final_ist = changed_tz.strftime('%H:00')
    ist_data_3.append([avg_points,final_ist])

# printing the top 5 and the least average number of comments posts
print('Top 5 hours for maximum comments on other posts in Indian Standard Time')
for row in ist_data_3[:5]:
        time_data = row[1]
        avg_points = row[0]
        print('{} IST: {:.2f} average comments per post'.format(time_data,avg_points))
print('Hour with least comments: {} IST with {:.2f} comments'.format(ist_data_3[-1][-1],ist_data_3[-1][0]))

Top 5 hours for maximum points on other posts in Indian Standard Time
23:00 IST: 62.53 average points per post
00:00 IST: 61.79 average points per post
01:00 IST: 60.54 average points per post
20:00 IST: 60.48 average points per post
05:00 IST: 60.01 average points per post
Hour with least points: 06:00 IST with 45.24 comments


Top 5 hours for maximum comments on other posts in Indian Standard Time
00:00 IST: 32.33 average comments per post
23:00 IST: 30.90 average comments per post
22:00 IST: 30.35 average comments per post
21:00 IST: 29.59 average comments per post
01:00 IST: 29.52 average comments per post
Hour with least comments: 16:00 IST with 21.36 comments


From above it can be seen that time from `13:00 to 14:00 EST` `(23:00 to 24:00 IST)` and `14:00 to 15:00 EST` `(00:00 to 01:00 IST)` are best for acheiving maximum comments and points on all the posts other than `Ask HN` and `Show HN` posts.

It is also worth noting that the percentage differences between the maximum and minimum values of average points and comments per hour per post are about `11%` and `17%`, which are quite less compared to the values we found above in the case of `Ask HN` and `Show HN` posts. This along with greater averaged values for comments and points for the other posts means that all the posts other than `Ask HN` and `Show HN` posts, are much more likely to engage a greater audience at almost all hours of the day